src/circuit/gadget/sinsemilla/merkle/chip.rs

```diff
@@ -0,0 +1,510 @@
+            let b_0 = left.value().map(|value| bitrange_subset(value, 240..250));
+
+            // b_1 = (bits 250..=254 of `left`)
+            // Constrain b_1 to 5 bits.
+            let b_1 = {
+                let b_1 = left
+                    .value()
+                    .map(|value| bitrange_subset(value, 250..L_ORCHARD_BASE));
+
+                config
+                    .sinsemilla_config
+                    .lookup_config_0
+                    .witness_short_check(layouter.namespace(|| "Constrain b_1 to 5 bits"), b_1, 5)?
+            };
+
+            // b_2 = (bits 0..=4 of `right`)
+            // Constrain b_2 to 5 bits.
+            let b_2 = {
+                let b_2 = right.value().map(|value| bitrange_subset(value, 0..5));
+
+                config
+                    .sinsemilla_config
+                    .lookup_config_1
+                    .witness_short_check(layouter.namespace(|| "Constrain b_2 to 5 bits"), b_2, 5)?
+            };
+
+            let b = {
+                let b = b_0
+                    .zip(b_1.value())
+                    .zip(b_2.value())
+                    .map(|((b_0, b_1), b_2)| {
+                        b_0 + b_1 * pallas::Base::from_u64(1 << 10)
+                            + b_2 * pallas::Base::from_u64(1 << 15)
+                    });
+                self.witness_message_piece(
+                    layouter.namespace(|| "Witness b = b_0||b_1||b_2||b_3"),
+                    b,
+                    2,
+                )?
+            };
+
+            (b_1, b_2, b)
+        };
+
+        let c = {
+            // `c = bits 5..=254 of `right`
+            let c = right
+                .value()
+                .map(|value| bitrange_subset(value, 5..L_ORCHARD_BASE));
+            self.witness_message_piece(layouter.namespace(|| "Witness c"), c, 25)?
+        };
+
+        let (point, zs) = self.hash_to_point(
+            layouter.namespace(|| format!("l_star {}", l_star)),
+            Q,
+            vec![a, b, c].into(),
+        )?;
+        let z1_a = zs[0][1];
+        let z1_b = zs[1][1];
+
+        // Check that the pieces have been decomposed properly.
+        {
+            layouter.assign_region(
+                || "Check piece decomposition",
+                |mut region| {
+                    // Set the fixed column `l_plus_1` to the current l_star + 1.
+                    let l_plus_1 = (l_star as u64) + 1;
+                    region.assign_fixed(
+                        || format!("l_plus_1 {}", l_plus_1),
+                        config.l_plus_1,
+                        0,
+                        || Ok(pallas::Base::from_u64(l_plus_1)),
+                    )?;
+
+                    // Offset 0
+                    // Copy and assign `a` at the correct position.
+                    copy(
+                        &mut region,
+                        || "copy a",
+                        config.advices[0],
```

Comment:
Maybe name these rather than repeating the indices from earlier?

Reply:
Instead of naming the columns, I have chosen to inline a table in the doc comments showing how they're being used in the custom gate:

```
The pieces and subpieces are arranged in the following configuration:
|  A_0  |  A_1  |  A_2  |  A_3  |  A_4  | l_plus_1 |
----------------------------------------------------
|   a   |   b   |   c   |  left | right |   l + 1  |
|  z1_a |  z1_b |  b_1  |  b_2  |       |          |
```